### ISSUE TITLE
`icefall-asr-gigaspeech-pruned-transducer-stateless2` pure accuracy when running with a streaming server

### ISSUE THREAD

**Issue by YuriiMytiai** [k2-fsa/sherpa#280]
Hello,
I tried to evaluate pretrained model `icefall-asr-gigaspeech-pruned-transducer-stateless2` for offline and streaming ASR and it looks like this model doesn't work properly with `streaming_pruned_transducer_statelessX` streaming server.
Offline decoding results look sane and WER on my data is ~17%.
The problem is that there are a lot of deletions (>60% deletions) when decoding with a streaming server. Even more, it looks like model predicts a lot of silence frames, because endpointing mechanism produces many short segments (2-3 sec long). 

The question is, is this expected behaviour for this model to work only in the offline mode? 
If so, what are the differences between regular and streaming models (according to the name, this model is dedicated streaming model: `icefall-asr-librispeech-pruned-stateless-streaming-conformer-rnnt4-2022-06-10`)?

Thank you!

**Comment by csukuangfj** [comment]
A streaming model can be only used with a streaming server.

By streaming we mean that we don't need to wait for the whole utterance to be spoken, we can decode as you are speaking.

For non-streaming models, you have to give it the whole utterance before it can start to decode, which also means the model has the full context of the utterance.

**Comment by YuriiMytiai** [comment]
Cool, thank you!
Where can I find a bit more technical info about differences in the models (or training approaches)?

**Comment by csukuangfj** [comment]
> Where can I find a bit more technical info about differences in the models (or training approaches)?

Please have a look at
- https://github.com/k2-fsa/icefall
- https://k2-fsa.github.io/icefall/ 

---

> I tried to decode with a dedicated streaming pretrained model (icefall-asr-librispeech-pruned-stateless-streaming-conformer-rnnt4-2022-06-10) in the offline and streaming modes

Please don't use a streaming model with an offline recognizer. 

>  I got worse results in terms of WER when decoding with streaming server

Could you post your commands for invoking the streaming server?

**Comment by YuriiMytiai** [comment]
> Could you post your commands for invoking the streaming server?
Sure, but nothing special at all - I used default decoder params:
`/home/ubuntu/sherpa/sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py --nn-model-filename /home/ubuntu/sherpa/icefall-asr-librispeech-pruned-stateless-streaming-conformer-rnnt4-2022-06-10/exp/cpu_jit-epoch-29-avg-6_torch-1.10.0.pt --bpe-model-filename /home/ubuntu/sherpa/icefall-asr-librispeech-pruned-stateless-streaming-conformer-rnnt4-2022-06-10/data/lang_bpe_500/bpe.model --decoding-method greedy_search --doc-root /home/ubuntu/sherpa/sherpa/bin/web --endpoint.rule1.must-contain-nonsilence false --endpoint.rule1.min-trailing-silence 0 --endpoint.rule1.min-utterance-length 10.0 --nn-pool-size 1 --port 6006`

**Comment by YuriiMytiai** [comment]
@csukuangfj I also have question about this 
<img width="941" alt="image" src="https://user-images.githubusercontent.com/34101338/212067553-0dd88fc1-d725-4ccc-860f-490e12eb73c5.png">
If I understand correctly, it means that I can take offline model (e.g. `icefall-asr-gigaspeech-pruned-transducer-stateless2`) and make it work properly for online ASR? If so, it's not that obvious how to do that from the PR mentioned in this message. Should I initialize with an offline model and train in the "streaming mode" for few more epochs with the same data? Or is it possible to make an offline model work online without training?

**Comment by csukuangfj** [comment]
>  it means that I can take offline model (e.g. icefall-asr-gigaspeech-pruned-transducer-stateless2) and make it work properly for online ASR?

No, an offline model is also known as a non-streaming model, while online ASR is also known as streaming ASR.

> Or is it possible to make an offline model work online without training?

No, as far as I know. (You may use a VAD to split an utterance into pieces and decode each piece with an offline model separately, but that is still not a streaming model)


If you want to train a streaming model, please see our doc
https://k2-fsa.github.io/icefall/recipes/index.html

If you want to download a pre-trained streaming model, please see also our doc at
https://k2-fsa.github.io/sherpa/cpp/pretrained_models/online_transducer.html

**Comment by csukuangfj** [comment]
> I tried to decode with a dedicated streaming pretrained model (icefall-asr-librispeech-pruned-stateless-streaming-conformer-rnnt4-2022-06-10)

How did you download the pre-trained model? I cannot find it in
https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/RESULTS.md
which contains all of our pre-trained models for the librispeech recipe.

**Comment by csukuangfj** [comment]
> --endpoint.rule1.must-contain-nonsilence false --endpoint.rule1.min-trailing-silence 0 --endpoint.rule1.min-utterance-length 10.0

I assume you know the meaning of the parameters for endpointing that you are using.
If not, please have a look at
https://k2-fsa.github.io/sherpa/python/streaming_asr/endpointing.html

(You can also have a look at https://k2-fsa.github.io/sherpa/ncnn/endpoint.html, which provides pictures for explanation)

**Comment by YuriiMytiai** [comment]
> How did you download the pre-trained model? I cannot find it in

I downloaded it from the Hugging Face:
`git clone https://huggingface.co/pkufool/icefall-asr-librispeech-pruned-stateless-streaming-conformer-rnnt4-2022-06-10`

**Comment by YuriiMytiai** [comment]
> No, an offline model is also known as a non-streaming model, while online ASR is also known as streaming ASR.
^ that is clear, I just used a different terminology=)

The point is that in the screenshot I posted the following is stated:
"If you want to adapt a non-streaming conformer model to be streaming, please refer to [this pull request](https://github.com/k2-fsa/icefall/pull/454).",
which I understand as "if you already have trained non-streaming conformer model and you want to use it in the streaming ASR, take a look at the PR and find info how you can adapt/modify your model in order to use it in the streaming ASR", but actually nothing about adapting/modifying non-streaming model for streaming ASR was found by me in the PR mentioned above.
